docassemble/AssemblyLine/custom_jinja_filters.py

```diff
@@ -200,11 +203,140 @@ def catchall_subquestion(value: Any, subquestion: str) -> DACatchAll:
     return value
 
 
+def _undefined_label(ud: Undefined, placeholder: Optional[str] = None) -> str:
+    """
+    Extract a friendly name from a Jinja2 Undefined.
+    Prefer the last missing attribute/key; fall back to the variable name;
+    then try parsing the hint; finally a generic label.
+
+    Args:
+        ud (Undefined): The Jinja2 Undefined object.
+        placeholder (str, optional): A generic label to use if no other name can be found.
+    Returns:
+        str: A friendly name extracted from the Undefined object.
+    """
+    # 1) Best case: attribute/key name (e.g., "signature" from users[0].signature)
+    name = getattr(ud, "_undefined_name", None)
+    if isinstance(name, str) and name:
+        # Note that name is already going to be just the last part of a dotted variable name,
+        # whic is perfect for a descriptive-ish placeholder.
+        return name
+
+    # 2) Sometimes the hint contains 'foo' or "foo"
+    hint = getattr(ud, "_undefined_hint", None)
+    if isinstance(hint, str):
+        m = re.search(r"""['"]([A-Za-z_][\w]*)['"]""", hint)
+        if m:
+            return m.group(1)
+
+    # 3) As a last resort, use a generic label
+    return str(placeholder)
+
+
+@pass_context
+def if_final(
+    context: Jinja2Context,
+    value: Any,
+    i: Optional[str] = None,
+    expected_values: Union[str, List[str]] = "final",
```

Comment:
```suggestion
    expected_i: Union[str, List[str]] = "final",
```

If we're hard coding this to just work with i, I'd change this variable name. It took me a while to realize it was talking about the values of `i`, and not the value that you're passing in `users[0].signature`, etc.

Reply:
Fair enough! I really can't think of another solid use case that isn't to make an easier syntax for our special case with `i` but if we come up with one, it should have a different name.